### PR TITLE
Sawa 152 154 156 POST PUT GET- Preferences 

### PR DESCRIPTION
**Pull request by fabiogiorgio**
I forgot to perform a last push in the previous PR.

### CODE REVIEW

File: ShowerShow/Control/UserPreferencesController.cs [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+        public UserPreferencesController(ILogger<UserPreferencesController> logger, IUserPrefencesService userPrefencesService, IUserService userService)
+        {
+            _logger = logger;
+            this._userPrefencesService = userPrefencesService;
+            this._userService = userService;
+
+        }
+
+        [Function("CreateUserPreferences")]
+        [OpenApiOperation(operationId: "CreatePreferences", tags: new[] { "Preferences" })]
```
Comment (egecnr) [inline ShowerShow/Control/UserPreferencesController.cs]: Add summary and description here. Also add summary and description to all your endpoints for more clearance.

File: ShowerShow/Control/UserPreferencesController.cs [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+        [OpenApiRequestBody("application/json", typeof(CreatePreferencesDTO), Description = "Id of the requested user ")]
+        [OpenApiResponseWithBody(statusCode: HttpStatusCode.Created, contentType: "application/json", bodyType: typeof(CreatePreferencesDTO), Description = "Successfully created the Preferences")]
+        public async Task <HttpResponseData> CreateUserPreferences([HttpTrigger(AuthorizationLevel.Anonymous, "post", Route = "user/{userId}/preferences")] HttpRequestData req)
```
Comment (egecnr) [inline ShowerShow/Control/UserPreferencesController.cs]: Where is authorization????

File: ShowerShow/Control/UserPreferencesController.cs [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+                responseData.StatusCode = HttpStatusCode.Created;
+
+                return responseData;
+            }
+        }
+        [Function("GetUserPreferenceById")]
+        [OpenApiOperation(operationId: "GetAllPreferences", tags: new[] { "Preferences" })]
+        [OpenApiParameter(name: "userId", In = ParameterLocation.Path, Required = true, Type = typeof(Guid), Description = "The user ID parameter")]
+        [OpenApiResponseWithBody(statusCode: HttpStatusCode.OK, contentType: "application/json", bodyType: typeof(List<Preferences>), Description = "Successfully retrieved the Preferences")]
+        public async Task<HttpResponseData> GetUserPreferenceById([HttpTrigger(AuthorizationLevel.Anonymous, "get", Route = "user/{userId:Guid}/preferences")] HttpRequestData req, Guid userId)
```
Comment (egecnr) [inline ShowerShow/Control/UserPreferencesController.cs]: Authorization

File: ShowerShow/Control/UserPreferencesController.cs [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+        [OpenApiRequestBody("application/json", typeof(UpdatePreferencesDTO), Description = "The user data to update.")]
+        [OpenApiResponseWithBody(statusCode: HttpStatusCode.OK, contentType: "application/json", bodyType: typeof(UpdatePreferencesDTO), Description = "The OK response with the updated user preference")]
+        public async Task<HttpResponseData> UpdatePreferenceById([HttpTrigger(AuthorizationLevel.Anonymous, "put", Route = "user/{userId:Guid}/preferences")] HttpRequestData req, Guid userId)
+        {
+            _logger.LogInformation($"Fetching the user by id {userId}");
+
+            string requestBody = await new StreamReader(req.Body).ReadToEndAsync();
+            UpdatePreferencesDTO updatePreferencesDTO = JsonConvert.DeserializeObject<UpdatePreferencesDTO>(requestBody);
+
+            if (await _userService.CheckIfUserExistAndActive(userId))
```
Comment (egecnr) [inline ShowerShow/Control/UserPreferencesController.cs]: You are checking if the user is active and then returning bad request.. Shouldnt you do the opposite?

File: ShowerShow/Model/Preferences.cs [hunk 5]
```diff
@@ -1,14 +1,37 @@
...
 namespace ShowerShow.Models
 {
-    public class Preferences
+    public partial class Preferences
     {
         public Guid Id { get; set; } = Guid.NewGuid();
         [JsonRequired]
         public Guid UserId { get; set; }
         public AvailableVoices SelectedVoice { get; set; } = AvailableVoices.Default_Male;
         public AvailableLanguages SelectedLanguage { get; set; } = AvailableLanguages.English;
```
Comment (egecnr) [inline ShowerShow/Model/Preferences.cs:13]: Voices are in blob storage. Check in with cosmin to understand it better. this class needs a tiny update

File: ShowerShow/Model/AvailableWaterTypes.cs [hunk 6]
```diff
@@ -0,0 +1,7 @@
+﻿namespace ShowerShow.Models
+{
+    public partial class Preferences
+    {
+        public enum AvailableWaterTypes { Liters, Gallons }
```
Comment (egecnr) [inline ShowerShow/Model/AvailableWaterTypes.cs:5]: If frontend agreed to these enums its all good

**Review by egecnr — DISMISSED** [review]
I have commented the parts that need update. Also try to improve the controller class by using try catch. I have updated user and userfriend controllers you cans ee them for reference when I make the prs.

Some of the controllers have the wrong logic (such as update preference) make sure to thoroughly test your endpoints before you make your pr

File: ShowerShow/Model/AvailableWaterTypes.cs (hunk 6 above)
Comment (egecnr) [reply ShowerShow/Model/AvailableWaterTypes.cs:5]: You can resolve this if you're in the same line with frontend

File: ShowerShow/Control/UserPreferencesController.cs [hunk 7]
```diff
@@ -0,0 +1,119 @@
...
+        [OpenApiResponseWithBody(statusCode: HttpStatusCode.Created, contentType: "application/json", bodyType: typeof(CreatePreferencesDTO), Description = "Successfully created the Preferences")]
+        public async Task <HttpResponseData> CreateUserPreferences([HttpTrigger(AuthorizationLevel.Anonymous, "post", Route = "user/{userId}/preferences")] HttpRequestData req)
+        {
+            _logger.LogInformation("Create a preference for one user");
+
+            string requestBody = await new StreamReader(req.Body).ReadToEndAsync();
+
+            CreatePreferencesDTO createPreferencesDTO = JsonConvert.DeserializeObject<CreatePreferencesDTO>(requestBody);
+
+            if (await _userService.CheckIfUserExistAndActive(createPreferencesDTO.UserId))
```
Comment (egecnr) [inline ShowerShow/Control/UserPreferencesController.cs]: For this case its correct you should only create pref if it doesnt exist this doesnt need to be fixed i think

Comment (fabiogiorgio) [reply ShowerShow/Control/UserPreferencesController.cs]: Ege said dont even have createpreference endpoint.only have a create method in repository

**Review by logoffro1 — APPROVED** [review]

**Pull request merged**